airbyte-cdk/python/unit_tests/sources/declarative/partition_routers/test_parent_state_stream.py

```diff
@@ -242,7 +242,10 @@ def _run_read(
         ]
     )
     logger = MagicMock()
-    return list(source.read(logger, config, catalog, state))
+    result = []
+    for m in source.read(logger, config, catalog, state):
+        result.append(copy.deepcopy(m))
```

Comment:
Why do we need a deepcopy here? I assume it is because we pull the state and the cursor modifies the state but it seems a bit dangerous that an Airbyte message is mutable...

Reply:
I believe it's because the same `stream_state` dictionary is referenced. It doesn't affect the sync since we don't buffer the output.

Reply:
After checking the code, it seems like [we return a new dict on every call](https://github.com/airbytehq/airbyte/blob/7b84e91f84ba65ec801e314a5e687680c71d1f62/airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/datetime_based_cursor.py#L107-L108). Hence, my hypothesis of "we re-use the same state" seems wrong. Do you think it is worth investigating more? It seems like a bad smell to me that we need to add that and I think there might be value in confirming if this is a change in behavior or it is just because of the new assertions we added and the behavior is the same as before.

Reply:
Yes, this change is needed because of the update I made to the test with intermediate states. I’ll investigate further.

Reply:
What is the outcome on this?